src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1682,10 +1682,13 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
         // UNDONE: Consider adding a secondary location pointing to the second method.
         private void ReportMethodSignatureCollision(DiagnosticBag diagnostics, SourceMemberMethodSymbol method1, SourceMemberMethodSymbol method2)
         {
-            // Partial methods are allowed to collide by signature.
-            if (method1.IsPartial && method2.IsPartial)
+            switch (method1, method2)
```

Comment:
>switch (method1, method2) [](start = 12, length = 25)

Does this `switch` statement represent a change in behavior from the original `if`?

Unless there is a change in behavior, we should stick with the `if` statement since it is significantly simpler.

Reply:
Yes, this fixes #43883. This solution ends up creating additional diagnostics in some error scenarios. I tried moving the checks for PartialMethodOnlyOneLatent/PartialMethodOnlyOneActual in here, but because partial method merging already happened in an earlier step, it causes us to miss diagnostics for multiple "actual" methods.

If we could do this at the same time or right before we merge partial methods, we might be able to get higher quality diagnostics in some scenarios when members have conflicting signatures:

1. if the methods are not 'partial', we report `ErrorCode.ERR_MemberAlreadyExists`
1. if we detect that both methods are partial implementations, we give `ErrorCode.ERR_PartialMethodOnlyOneActual`
1. if we detect that both methods are partial definitions, we give `ErrorCode.ERR_PartialMethodOnlyOneLatent`

It might also be possible to fix this by modifying the partial method signature comparer, that groups declarations that might be part of the same partial method, to be aware of RefKind differences.

Reply:
> This solution ends up creating additional diagnostics in some error scenarios.

That sounds ok. We can revisit this later if necessary.

---
In reply to: [418977822](https://github.com/dotnet/roslyn/pull/43582#discussion_r418977822) [](ancestors = 418977822)